Cython/Compiler/Nodes.py

```diff
@@ -8937,6 +8937,129 @@ def generate_execution_code(self, code):
         code.put_ensure_gil(declare_gilstate=False)
 
 
+class CriticalSectionStatNode(TryFinallyStatNode):
+    """
+    Represents a freethreading Python critical section.
+    In non-freethreading Python, this is a no-op.
+
+    args    list of ExprNode    1 or 2 elements, must be object
+    """
+
+    child_attrs = ["args"] + TryFinallyStatNode.child_attrs
+
+    # loosely follow the naming convention that Python itself follows
+    DEFAULT_VARIABLE_NAME = "__pyx_cs"
+
+    var_type = None
+    state_temp = None
+
+    def __init__(self, pos, /, args, body, **kwds):
+        if len(args) > 1:
+            self.var_type = PyrexTypes.c_py_critical_section2_type
+        else:
+            self.var_type = PyrexTypes.c_py_critical_section_type
+
+        self.create_state_temp_if_needed(pos, args, body)
+
+        TryFinallyStatNode.__init__(
+            self, pos,
+            args=args,
+            body=body,
+            **kwds,
+            finally_clause=CriticalSectionExitNode(
+                pos, len=len(args), state_temp=self.state_temp))
+
+    def create_state_temp_if_needed(self, pos, args, body):
+        from .ParseTreeTransforms import YieldNodeCollector
+        collector = YieldNodeCollector()
+        collector.visitchildren(body)
+        if not collector.yields:
+            return
+
+        from . import ExprNodes
+        self.state_temp = ExprNodes.TempNode(pos, self.var_type)
+
+    def analyse_declarations(self, env):
+        for arg in self.args:
+            arg.analyse_declarations(env)
+        return super().analyse_declarations(env)
+
+    def analyse_expressions(self, env):
+        env.use_utility_code(
+            UtilityCode.load_cached("CriticalSections", "ModuleSetupCode.c"))
+
+        for n in range(len(self.args)):
+            # Coerce to temp because it's a bit of a disaster if the argument is destroyed
+            # while we're working on it, and the Python critical section implementation
+            # doesn't ensure this.
+            # TODO - we could potentially be a bit smarter about this, and avoid
+            # it for local variables that we know are never re-assigned.
+            self.args[n] = self.args[n].analyse_expressions(env).coerce_to_temp(env)
+            # Note - deliberately no coercion to Python object.
+            # Critical sections only really make sense on a specific known Python object,
+            # so using them on coerced Python objects is very unlikely to make sense.
+            if not self.args[n].type.is_pyobject:
+                error(
+                    self.pos,
+                    "Arguments to cython.critical_section must be Python objects."
+                )
```

Comment:
- why loop over `range(len(…))` when you still need to access the list items in the loop?
- error on `arg.pos`, not `self.pos`
```suggestion
        for i, arg in enumerate(self.args):
            # Coerce to temp because it's a bit of a disaster if the argument is destroyed
            # while we're working on it, and the Python critical section implementation
            # doesn't ensure this.
            # TODO - we could potentially be a bit smarter about this, and avoid
            # it for local variables that we know are never re-assigned.
            arg = arg.analyse_expressions(env).coerce_to_temp(env)
            # Note - deliberately no coercion to Python object.
            # Critical sections only really make sense on a specific known Python object,
            # so using them on coerced Python objects is very unlikely to make sense.
            if not arg.type.is_pyobject:
                error(
                    arg.pos,
                    "Arguments to cython.critical_section must be Python objects."
                )
            self.args[i] = arg
```